### PR TITLE
Migrate to next.js

### PR DESCRIPTION
**Pull request by ryanjduffy**
## Issue

Migrating to next.js will improve our deployment story on vercel and unlock hot reloading and server-side rendering

## Resolution

This touches a lot but the highlights are:

* Config updates incl dropping babel and webpack config in favor of a next.config.js. Updating jest config to transpile correctly
* Few random tranpilation fixes (removing a couple comments that were breaking a `return`, removing `*/` in a css file that broke minification)
* Added `/pages` directory to serve our pages along with handlers for the modals
* Moved the `extendReducer` logic around so it was tied to components rather than module loading
* Several guards for `window` for SSR

### CODE REVIEW

File: next.config.js [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    config.entry = () => {
+      return entry().then(e => ({
+        ...e,
+        parserWorker: "./src/devtools/client/debugger/src/workers/parser/worker",
+        searchWorker: "./src/devtools/client/debugger/src/workers/search/worker",
+      }));
+    };
+
+    // handles build error from webpack/runtime/compat
+    if (isServer) {
```
Comment (jasonLaster) [inline next.config.js:14]: not sure if we need this if we're relying on vercel to build

File: next.config.js [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    //       options: {
+    //         importLoaders: 1,
+    //         url: {
+    //           filter: (url, resourcePath) => resourcePath.endsWith("/src/image/image.css"),
+    //         },
+    //       },
+    //     },
+    //     "postcss-loader",
+    //   ],
+    // });
```
Comment (jasonLaster) [inline next.config.js:74]: we can delete this

File: src/devtools/client/debugger/packages/devtools-reps/src/object-inspector/components/ObjectInspector.tsx [hunk 3]
```diff
@@ -19,7 +19,7 @@ const {
   nodeHasSetter,
 } = Utils.node;
 
-require("./ObjectInspector.css");
+// require("./ObjectInspector.css");
```
Comment (jasonLaster) [inline src/devtools/client/debugger/packages/devtools-reps/src/object-inspector/components/ObjectInspector.tsx:22]: we can delete these with project search in vscode

File: src/devtools/client/debugger/src/components/App.css [hunk 4]
```diff
@@ -1,8 +1,8 @@
 /* This Source Code Form is subject to the terms of the Mozilla Public
- * License, v. 2.0. If a copy of the MPL was not distributed with this
- * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */
+* License, v. 2.0. If a copy of the MPL was not distributed with this
+* file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */
 
-button {
+.button {
```
Comment (jasonLaster) [inline src/devtools/client/debugger/src/components/App.css:5]: why do we need to do this?

File: src/devtools/client/debugger/src/components/App.js [hunk 5]
```diff
@@ -34,13 +34,13 @@ const { appinfo } = Services;
...
 
-import "./App.css";
+// import "./App.css";
 
-import "devtools/client/themes/variables.css";
-import "devtools/client/themes/webconsole.css";
-import "devtools/client/themes/common.css";
+// import "devtools/client/themes/variables.css";
+// import "devtools/client/themes/webconsole.css";
+// import "devtools/client/themes/common.css";
```
Comment (jasonLaster) [inline src/devtools/client/debugger/src/components/App.js:41]: we can delete these with project search (find and replace in vscode)

File: src/ui/components/shared/WorkspaceSettingsModal/WorkspaceSettingsModal.tsx [hunk 6]
```diff
@@ -194,19 +193,20 @@ const settings: Settings<
...
 
 function WorkspaceSettingsModal({ workspaceId, ...rest }: PropsFromRedux) {
-  const history = useHistory();
-  const { pathname } = useLocation();
-  const match = matchPath<{ workspaceId: string }>(pathname, {
-    path: "/team/:workspaceId/settings/billing",
-  });
+  // const history = useHistory();
+  // const { pathname } = useLocation();
+  const match = false;
```
Comment (jasonLaster) [inline src/ui/components/shared/WorkspaceSettingsModal/WorkspaceSettingsModal.tsx:198]: should we remove match too?

File: src/ui/components/shared/WorkspaceSettingsModal/WorkspaceMember.tsx [hunk 7]
```diff
@@ -8,7 +8,7 @@ import * as actions from "ui/actions/app";
 import MaterialIcon from "../MaterialIcon";
 import { AvatarImage } from "ui/components/Avatar";
 
-import "./WorkspaceMember.css";
+// import "./WorkspaceMember.css";
```
Comment (jasonLaster) [inline src/ui/components/shared/WorkspaceSettingsModal/WorkspaceMember.tsx:11]: it'd be nice to have a separate PR that consolidates the css imports

File: src/ui/components/shared/WorkspaceSettingsModal/WorkspaceSettingsModal.tsx [hunk 8]
```diff
@@ -194,19 +193,20 @@ const settings: Settings<
...
+  // });
   const [defaultTab, setDefaultTab] = useState<string>("Team Members");
   const { members } = hooks.useGetWorkspaceMembers(workspaceId!);
   const { userId: localUserId } = hooks.useGetUserId();
 
   useEffect(() => {
     if (match) {
       setDefaultTab("Billing");
-      history.replace("/");
+      // history.replace("/");
```
Comment (jasonLaster) [inline src/ui/components/shared/WorkspaceSettingsModal/WorkspaceSettingsModal.tsx:209]: lets actually delete this

File: src/ui/utils/apolloClient.tsx [hunk 9]
```diff
@@ -145,7 +145,7 @@ function createHttpLink(token: string | undefined) {
     headers["Authorization"] = `Bearer ${token}`;
   }
 
-  const uri = process.env.API_URL;
+  const uri = process.env.NEXT_PUBLIC_API_URL;
```
Comment (jasonLaster) [inline src/ui/utils/apolloClient.tsx:148]: nice

File: src/ui/utils/intercom.ts [hunk 10]
```diff
@@ -7,10 +8,14 @@ declare global {
...
+  requiresWindow(() => {
+    // window.Intercom("shutdown");
+    logout({ returnTo: window.location.origin });
+  });
 }
 
 export function bootIntercom(data: any) {
-  window.Intercom("boot", { app_id: APP_ID, ...data });
+  requiresWindow(() => {
+    // window.Intercom("boot", { app_id: APP_ID, ...data });
```
Comment (jasonLaster) [inline src/ui/utils/intercom.ts:19]: should we bring this back?

File: src/ui/utils/prefs.ts [hunk 11]
```diff
@@ -2,16 +2,19 @@ import { PrefsHelper } from "devtools/client/shared/prefs";
...
+// function getUrlPrefs() {
+//   return usesWindow(win => {
+//     if (!win) return {};
 
-const urlPrefs = getUrlPrefs();
+//     const url = new URL(window.location.href);
+//     const urlPrefs = url.searchParams.get("prefs") || "";
+//     return Object.fromEntries(urlPrefs.split(",").map(pref => pref.split(":")));
+//   });
+// }
```
Comment (jasonLaster) [inline src/ui/utils/prefs.ts:17]: what do we want to do here? Ultimately we might not need this too

**Review by jasonLaster — COMMENTED** [review]
Couple of small comments

File: next.config.js (hunk 1 above)
Comment (ryanjduffy) [reply next.config.js:14]: I don't know if it's required. I'm following a reference in the (now linked) issue that said they used that to avoid extra size in their client bundle.

File: src/devtools/client/debugger/src/components/App.css (hunk 4 above)
Comment (ryanjduffy) [reply src/devtools/client/debugger/src/components/App.css:5]: we don't. it was from my CSS fix exploration. i'll revert

File: src/ui/utils/prefs.ts (hunk 11 above)
Comment (ryanjduffy) [reply src/ui/utils/prefs.ts:17]: I realized after I "fixed" it that it wasn't used right now. No objection from me to removing it if you agree.

File: src/ui/utils/intercom.ts (hunk 10 above)
Comment (ryanjduffy) [reply src/ui/utils/intercom.ts]: Done!

File: src/ui/utils/prefs.ts (hunk 11 above)
Comment (jasonLaster) [reply src/ui/utils/prefs.ts:17]: kill it

File: src/ui/utils/routeListener.tsx [hunk 12]
```diff
@@ -1,17 +1,21 @@
...
-  const history = useHistory();
+  const router = useRouter();
   if (!listenerInstalled) {
-    history.listen((location, action) => {
+    currentRoute = window.location.pathname;
+
+    router.events.on("routeChangeComplete", () => {
       const newRoute = location.pathname;
-      if (currentRoute.startsWith("/recording/") && newRoute === "/") {
+      if (currentRoute && currentRoute.startsWith("/recording/") && newRoute === "/") {
```
Comment (jcmorrow) [inline src/ui/utils/routeListener.tsx:13]: `currentRoute?.startsWith`?

File: src/ui/components/shared/UserSettingsModal/UserSettingsModal.tsx [hunk 13]
```diff
@@ -231,19 +231,19 @@ export function UserSettingsModal(props: PropsFromRedux) {
...
-  useEffect(() => {
-    if (match) {
-      history.replace("/");
-    }
-  }, []);
+  // useEffect(() => {
+  //   if (match) {
+  //     history.replace("/");
+  //   }
+  // }, []);
```
Comment (jasonLaster) [inline src/ui/components/shared/UserSettingsModal/UserSettingsModal.tsx:238]: lets delete

File: src/ui/components/shared/UserSettingsModal/UserSettingsModal.tsx [hunk 14]
```diff
@@ -205,11 +204,12 @@ const getSettings = (internal: boolean): Settings<SettingsTabTitle, UserSettings
...
-  const { pathname } = useLocation();
-  const match = matchPath(pathname, {
-    path: "/settings/preferences",
-  });
+  // const history = useHistory();
+  // const { pathname } = useLocation();
+  const match = false;
+  // matchPath(pathname, {
+  //   path: "/settings/preferences",
+  // });
```
Comment (jasonLaster) [inline src/ui/components/shared/UserSettingsModal/UserSettingsModal.tsx:212]: 🔫

File: src/ui/components/Header/Header.module.css [hunk 15]
```diff
@@ -44,11 +38,14 @@
...
 
-#header .links {
+.header .links {
   display: flex;
   align-items: center;
 }
 
+<<<<<<< HEAD:src/ui/components/Header/Header.module.css
+.header .links .avatars {
+=======
```
Comment (jasonLaster) [inline src/ui/components/Header/Header.module.css:48]: 😉

File: src/ui/components/Account/index.tsx [hunk 16]
```diff
@@ -15,6 +13,23 @@ import {
   OnboardingHeader,
   OnboardingModalContainer,
 } from "../shared/Onboarding/index";
+import { useRouter } from "next/dist/client/router";
+
+/**
+ * Create a (host relative) URL with the given parameters. Used for replacing
+ * legacy routes with their current equivalents.
+ */
+function createReplayURL(currentParams: URLSearchParams) {
```
Comment (jasonLaster) [inline src/ui/components/Account/index.tsx:22]: 👍

File: src/image/image.css [hunk 17]
```diff
@@ -1,178 +1,178 @@
 .img.webpack {
-  background-image: url(~devtools/client/debugger/images/sources/webpack.svg);
+  background-image: url(~devtools/client/debugger/images/sources/webpack.svg) !important;
```
Comment (jasonLaster) [inline src/image/image.css:2]: as a reference, why do we need this?

File: src/devtools/client/inspector/computed/actions/index.ts [hunk 18]
```diff
@@ -51,14 +51,14 @@ export function setComputedPropertyExpanded(
...
 
-const outputParser = new OutputParser(document, CSSProperties);
-
 async function createComputedProperties(
   elementStyle: ElementStyle
 ): Promise<ComputedPropertyState[]> {
   const computed = await elementStyle.element.getComputedStyle();
   if (!computed) return [];
 
+  const outputParser = new OutputParser(document, CSSProperties);
```
Comment (jasonLaster) [inline src/devtools/client/inspector/computed/actions/index.ts:60]: Nice

**Review by jasonLaster — APPROVED** [review]
one thing

File: src/base.css [hunk 19]
```diff
@@ -18,7 +18,8 @@ body {
   border: 2px solid black !important;
 }
 
-#app {
+#app,
```
Comment (jcmorrow) [inline src/base.css:21]: Will we even have a `#app` anymore? If not, should we remove these lines?

**Review by jcmorrow — COMMENTED** [review]
Exciting times! Left some comments, feel free to ping me or push back on any of the suggestions! (except for the merge conflicts, should probably get that out of there 😉 )

File: src/image/image.css (hunk 17 above)
Comment (ryanjduffy) [reply src/image/image.css:2]: There's an import order problem where (some, at least) of these weren't taking effect because they used to be. imported after another rule that had the same precedence. This was my "let's just ship this already" bail out for that problem. Follow up required.

File: src/ui/components/Header/Header.module.css [hunk 20]
```diff
@@ -23,14 +24,7 @@
...
 
-#header .header-left {
-  display: flex;
-  flex-direction: row;
-  align-items: center;
-  position: relative;
-}
-
-#header .title-container {
+.header .title-container {
```
[the remark this answers is not in the document]
Comment (ryanjduffy) [reply src/ui/components/Header/Header.module.css:27]: We can. I'm going to punt on this for now but would be a great candidate for a first follow up to get a good archetype for our CSS module migration.

File: src/ui/components/Header/Header.tsx [hunk 21]
```diff
@@ -29,7 +32,7 @@ function Avatars({ recordingId }: { recordingId: RecordingId | null }) {
   }
 
   return (
-    <div className="avatars">
+    <div className={css.avatars}>
```
[the remark this answers is not in the document]
Comment (ryanjduffy) [reply src/ui/components/Header/Header.tsx:35]: My pref is this form. If you pass values to `classnames` and they aren't found, a global class is returned which could introduce some oddities. Better to not, imo.

File: src/ui/components/shared/SharingModal/SharingModal.tsx [hunk 22]
```diff
@@ -12,7 +12,8 @@ import MaterialIcon from "../MaterialIcon";
 import PrivacyDropdown from "./PrivacyDropdown";
 
 function SharingModalWrapper(props: PropsFromRedux) {
-  const { recordingId } = props.modalOptions!;
+  const opts = props.modalOptions;
+  const recordingId = opts && "recordingId" in opts ? opts.recordingId : "";
```
[the remark this answers is not in the document]
Comment (ryanjduffy) [reply src/ui/components/shared/SharingModal/SharingModal.tsx:16]: Typescript is ornery here. It'll complain because opts either has `recordingId` or `view` at the moment. I I think the fix is to build the action differently.

File: src/ui/components/shared/UserSettingsModal/UserSettingsModal.tsx (hunk 13 above)
[the remark this answers is not in the document]
Comment (ryanjduffy) [reply src/ui/components/shared/UserSettingsModal/UserSettingsModal.tsx:238]: done

File: src/ui/hooks/recordings.ts [hunk 23]
```diff
@@ -104,8 +104,8 @@ const GET_MY_RECORDINGS = gql`
 `;
 
 export function useGetRecordingId() {
-  const { recordingId } = useParams<{ recordingId: string }>();
-  return recordingId;
+  const { id } = useRouter().query;
+  return Array.isArray(id) ? id[0] : id!;
```
[the remark this answers is not in the document]
Comment (ryanjduffy) [reply src/ui/hooks/recordings.ts:108]: We'd never see it but the next.js typings allow for it. I think one of two ways:

`/path/:id/asdf?id=` - would give you two `id` values, or
`/path/[...id]` would give you N `id` values

Neither are applicable here but ... typescript ...

File: src/ui/utils/environment.ts [hunk 24]
```diff
@@ -1,5 +1,6 @@
 import { MockedResponse } from "@apollo/client/testing";
-import { matchPath } from "react-router-dom";
+import _ from "lodash";
```
[the remark this answers is not in the document]
Comment (ryanjduffy) [reply src/ui/utils/environment.ts:2]: Nope. I'm sure it was VSCode being "helpful"

File: src/base.css (hunk 19 above)
Comment (ryanjduffy) [reply src/base.css:21]: works for me! I was being conservative initially until everything was moved. Cleaning now
